lib/web_ui/lib/src/engine/skwasm/skwasm_impl/codecs.dart

```diff
@@ -0,0 +1,33 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:js_interop';
+
+import 'package:ui/src/engine.dart';
+import 'package:ui/src/engine/skwasm/skwasm_impl.dart';
+import 'package:ui/ui.dart' as ui;
+
+class SkwasmImageDecoder extends BrowserImageDecoder {
+  SkwasmImageDecoder({
+    required super.contentType,
+    required super.dataSource,
+    required super.debugSource,
+  });
+
+  @override
+  ui.Image generateImageFromVideoFrame(VideoFrame frame) {
+    final int width = frame.codedWidth.toInt();
+    final int height = frame.codedHeight.toInt();
+    final SkwasmSurface surface = (renderer as SkwasmRenderer).surface;
+    final int videoFrameId = surface.acquireObjectId();
+    skwasmInstance.skwasmRegisterObject(videoFrameId.toJS, frame as JSAny);
+    skwasmInstance.skwasmTransferObjectToThread(videoFrameId.toJS, surface.threadId.toJS);
+    return SkwasmImage(imageCreateFromVideoFrame(
```

Comment:
Is there a potential race between the transfer and SkImage initialization?

Reply:
No, there isn't, although I could see why you might think so. It is true that the `SkImage` is being created synchronously on the UI thread, but it's the image's *generator* that actually attempts to grab the video frame by its ID and convert it to a texture, and that generator is always run on the web worker right before the image is being either drawn to the actual graphics surface or being rasterized to pixels. Any call that would draw or rasterize the image is sent to the web worker via a postMessage, and the postMessage queue is serial, so those calls are guaranteed to happen *after* the postMessage that transfers the video frame.